### PR TITLE
heffte: fix conflict directive (missing '^')

### PR DESCRIPTION
**Pull request by aweits**
discovered during testing #19501 
@mkstoyanov

### CODE REVIEW

**Review by alalazo — APPROVED** [review]

File: var/spack/repos/builtin/packages/heffte/package.py [hunk 1]
```diff
@@ -40,7 +40,7 @@ class Heffte(CMakePackage):
 
     conflicts('~fftw', when='~mkl~cuda')  # requires at least one backend
     conflicts('+fftw', when='+mkl@:1.0')  # old API supports at most one CPU backend
-    conflicts('openmpi~cuda', when='+cuda')  # +cuda requires CUDA enabled OpenMPI
+    conflicts('^openmpi~cuda', when='+cuda')  # +cuda requires CUDA enabled OpenMPI
```
Comment (mkstoyanov) [inline var/spack/repos/builtin/packages/heffte/package.py:43]: I did not know that, thanks for catching it.

**Review by mkstoyanov — APPROVED** [review]

File: var/spack/repos/builtin/packages/heffte/package.py (hunk 1 above)
Comment (balay) [reply var/spack/repos/builtin/packages/heffte/package.py:43]: Is this a real requirement of heffte? And one has to use the latest - not default openmpi for openmpi+cuda?

And one can't use heffte with mpich? i.e is this a mandatory requirement?
```
+    depends_on('openmpi@4.0.5 +cuda', when='+cuda +heffte')
```

This change breaks current xsdk test [for heffte..]

Comment (mkstoyanov) [reply var/spack/repos/builtin/packages/heffte/package.py:43]: heFFTe is using CUDA-Aware MPI communication, basically we can pass-in an array that sits in the GPU and MPI will "know" what to do to send it directly.

I don't know if mpich has limitations when working with CUDA, but there was no CUDA variant for mpich.

OpenMPI default (3.1.6) works with CUDA, but there is a conflict with hwlock:
```
spack spec openmpi+cuda (fails)
spack spec openmpi+cuda ^hwloc@1.11.11+cuda  (works fine)
```

Spectrum MPI on Summit is CUDA-Aware, although you may need some extra flags at runtime (compiling is fine).

Without CUDA, heFFTe works with mpich.

Comment (balay) [reply var/spack/repos/builtin/packages/heffte/package.py:43]: Would the above conflict prevent both mpich and spectrum-mpi from getting used? [when cuda is enabled]

Presumably you want an error if mpich is selected - but not when spectrum-mpi is selected.

@luszczek is testing on summit - don't know if this test includes cuda/spectrum-mpi/heffte.

Comment (mkstoyanov) [reply var/spack/repos/builtin/packages/heffte/package.py:43]: I don't know how to express that. As far as I know, there is no way to specify "or" dependence in spack.

Comment (luszczek) [reply var/spack/repos/builtin/packages/heffte/package.py:43]: @mkstoyanov , what are you trying to express? `(openmpi without cuda WHEN cuda variant) OR (???)`

Comment (mkstoyanov) [reply var/spack/repos/builtin/packages/heffte/package.py:43]: I am saying that heFFTe requires CUDA-Aware MPI for the variant +cuda.

An alternative way to say this is:
```
conflicts('^mpich', when = '+cuda')
conflicts('^openmpi~cuda', when='+cuda')
```
That should allow spectrum MPI to be added, but I am not sure about other implementations of MPI.

**Pull request merged**